### PR TITLE
Add resource watcher to log metadata

### PR DESCRIPTION
**Pull request by zhenghuiwang**
Fixes https://github.com/kubeflow/metadata/issues/97

See [README.md](https://github.com/kubeflow/metadata/pull/126/files#diff-b274fa8b4e136bde65013dd08aeecd2e) / [watcher.go](https://github.com/kubeflow/metadata/pull/126/files#diff-a14801b193ef46f3422f0a8977e7942a) /[main.go](https://github.com/kubeflow/metadata/pull/126/files#diff-7414adc256a7ffe0fa4dec0be4c03c3f)

This PR add a POC for watcher: get a list of resources types and log newly created objects into metadata store.

**TODOs:**
- [x] Refine the Metadata type definition to capture more information such as creation time & URI.
- [x] Containerize the watcher and deploy it as a pod.
- [x] Use newly deployed gRPC MLMD server to log metadata. The openapi generated Metadata service go client doesn't compile. I have to manually change the generated code. (Ideally, we don't need to check in the go client code) I would wait for the new gRPC server for v0.7 release.
- [x] Integration tests.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/metadata/126)

### CODE REVIEW

**Comment by zhenghuiwang** [comment]
It is ready for review!
/assign @jlewi 
/cc @hougangliu @dreamryx

File: watcher/main/main.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+		os.Exit(1) // second signal. Exit directly.
+	}()
+	return stop
+}
+
+func init() {
+	flag.StringVar(&kubeconfig, "kubeconfig", "", "Path to a kubeconfig. Only required if out-of-cluster.")
+	flag.StringVar(&masterURL, "master", "", "The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
+	flag.StringVar(&metadataServiceURL, "metadata_service", "localhost:8080", "The address of the Kubeflow Metadata GRPC service.")
+	flag.StringVar(&resourcelist, "resourcelist", "", "Path to a JSON file with a list of Kubernets GroupVersionKind to be watched.")
```
Comment (richardsliu) [inline watcher/main/main.go]: Typo for Kubernetes. Also mark which flags are required and which are optional.

File: watcher/handlers/metalogger.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+	if err != nil {
+		return l, fmt.Errorf("failed to create artifact type: err = %v; request = %v; response = %v", err, request, resp)
+	}
+	l.typeID = resp.GetTypeId()
+	return l, nil
+}
+
+// MetadataArtifactType returns the metadata artifact type for the MetaLogger's GroupVerionKind
+func (l *MetaLogger) MetadataArtifactType() string {
+	return fmt.Sprintf("kubeflow.org/%s/%s", l.resource.GroupKind(), l.resource.Version)
```
Comment (richardsliu) [inline watcher/handlers/metalogger.go]: Are we only watching KF resources?

File: watcher/handlers/metalogger.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+			return fmt.Errorf("error decoding object tombstone, invalid type")
+		}
+	}
+	exists, err := l.ifExists(object.GetUID())
+	if err != nil {
+		klog.Errorf("failed to check if object of type %s exisits: %s", l.MetadataArtifactType(), err)
+		return err
+	}
+	if exists {
+		klog.Infof("Hanlded addEvent for %s. Object already exists with UID = %s, name = %s.\n", l.MetadataArtifactType(), object.GetUID(), object.GetName())
```
Comment (richardsliu) [inline watcher/handlers/metalogger.go]: Typo - handled

File: watcher/handlers/metalogger.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+	}
+	request := storepb.PutArtifactsRequest{
+		Artifacts: []*mlpb.Artifact{&artifact},
+	}
+	resp, err := l.kfmdClient.PutArtifacts(context.Background(), &request)
+	if err != nil {
+		klog.Errorf("failed to log metadata for %s: err = %s, request = %v, resp = %v", l.MetadataArtifactType(), err, request, resp)
+		return err
+	}
+	klog.Infof("Hanlded addEvent for %s.\n", l.MetadataArtifactType())
```
Comment (richardsliu) [inline watcher/handlers/metalogger.go]: same

File: watcher/handlers/metalogger.go (hunk 2 above)
Comment (zhenghuiwang) [reply watcher/handlers/metalogger.go]: No. The resource to be watched is configured by the `resourcelist` flag. By default, they are `pod` and `service`

this type is the metadata artifact type. I'm add a kubeflow prefix here to distinguish it from others.

**Pull request merged**